runtime/compiler/codegen/J9CodeGenerator.cpp

```diff
@@ -3917,6 +3918,118 @@ J9::CodeGenerator::collectSymRefs(
+    *       For virtual dispatch sequence, we know that this is the safe check but in case of interface call, classes implementing
+    *       that interface would have different size of VTable. This makes executing above check unsafe when VTable of the class of
+    *       the receiver object is smaller, effectively making reference in n3n to pointing to a garbage location which might lead 
+    *       to a segmentation fault if the reference in not memory mapped or if bychance it contains J9Method  pointer of same inlined
+    *       method then it will execute a code which should not be executed.
+    *       For this kind of Virtual guards which are not nop'd we need to add a range check to make sure the address we are going to
+    *       access is pointing to a valid location in VTable. There are mainly two ways we can add this range check test. First one is
+    *       during the conception of the virtual guard. There are many downsides of doing so especially when other optimizations which
+    *       can moved guards around (for example loop versioned, virtualguard head merger, etc) needs to make sure to move range check 
+    *       test around as well. Other way is to scan for this type of guards after optimization is finished like here in CodeGen Phase
+    *       and add a range check test here.
+    *       At the end of this function, we would have following code around them method test.
+    *       BBStart <block_X>    
+    *       ...      
+    *       ifacmple goto nXXn
+    *          aloadi <offset of VTableHeader.size from J9Class*>
+    *             aload <vft>
+    *          aconst <Index of the inlined method in VTable of class of inlined method>
+    *       BBEnd <block_X>
+    * 
+    *       BBStart <block_Y>
+    *       ifacmpne goto -> nXXn 
+    *          aloadi <offset of inlined method in VTable>
+    *             aload <vft>
+    *          aconst <J9Method of inlined method>
+    *       BBEnd <block_Y>
+    *               
+    *       
+    */  
+void
+J9::CodeGenerator::fixUpProfiledInterfaceGuardTest()
+   {
+   TR::Compilation *comp = self()->comp();
+   TR::CFG * cfg = comp->getFlowGraph();
+   TR::NodeChecklist checklist(comp);
+   for (TR::AllBlockIterator iter(cfg, comp); iter.currentBlock() != NULL; ++iter)
+      {
+      TR::Block *block = iter.currentBlock();
+      TR::TreeTop *treeTop = block->getLastRealTreeTop();
+      TR::Node *node = treeTop->getNode();
+      if (node->getOpCode().isIf() && node->isTheVirtualGuardForAGuardedInlinedCall() && !checklist.contains(node))
+         {
+         TR_VirtualGuard *vg = comp->findVirtualGuardInfo(node);
+         // Mainly we need to make sure that virtual guard which performs the TR_MethodTest and can be NOP'd are needed the range check.
+         if (vg && vg->getTestType() == TR_MethodTest && 
+            !(comp->performVirtualGuardNOPing() && (node->isNopableInlineGuard() || comp->isVirtualGuardNOPingRequired(vg))))
```

Comment:
@andrewcraik @jdmpapin  Combining query from CodeGen where it generates the VGNOP. Basically if VGNOP is supported and node or guard requires NOPing then we won't add range check test.

Reply:
I think this is ok - @jdmpapin concerns?

Reply:
Something like this is probably the best we'll get without some mechanism to get codegen to commit in advance to using `vgnop`. I think it's good for now